gyronimo/dynamics/cartesian_boris.cc

```diff
@@ -29,91 +29,49 @@
-  // 							" Efield cannot be nullptr.",
-  // 1);
-  if (!Bfield)
-    error(__func__, __FILE__, __LINE__, " Bfield cannot be nullptr.", 1);
-
-  // test if fields are cartesian
+  if (!B) error(__func__, __FILE__, __LINE__, "no magnetic field.", 1);
+  if (E && B->metric() != E->metric())
+    error(__func__, __FILE__, __LINE__, "mismatched E/B coordinates.", 1);
   if (!metric_)
-    error(
-        __func__, __FILE__, __LINE__,
-        " Bfield must be in cartesian coordinates.", 1);
-  if (Efield) {
-    const metric_covariant* Emetric = Efield->metric();
-    if (Emetric != metric_)
-      error(
-          __func__, __FILE__, __LINE__,
-          " Efield must be in cartesian coordinates.", 1);
-  }
-
-  // test if normalization factors are consistent
-  if (Efield) {
-    double E_m_factor = Efield->m_factor();
-    double B_m_factor = Bfield->m_factor();
-    if (std::abs(1.0 - E_m_factor / (Vref_ * B_m_factor)) > 1.0e-14)
-      error(
-          __func__, __FILE__, __LINE__, " inconsistent Efield and Bfield.", 1);
-  }
+    error(__func__, __FILE__, __LINE__, "field has no metric_cartesian.", 1);
 }
 
 //! Performs a time step `dt` update to the state `in` and returns the result.
 cartesian_boris::state cartesian_boris::do_step(
-    const cartesian_boris::state& in, const double& time,
-    const double& dt) const {
-  // extract position and velocity from state
-  IR3 x_old = {in[0], in[1], in[2]};
-  IR3 v_old = {in[3], in[4], in[5]};
-
-  // calculate fields
-  IR3 Efield = {0.0, 0.0, 0.0};
-  if (electric_field_) {
-    double Etime = time * iEfield_time_factor_;
-    Efield = electric_field_->contravariant(x_old, Etime);
-  }
-  double Btime = time * iBfield_time_factor_;
-  double Bmag = magnetic_field_->magnitude(x_old, Btime);
-  IR3 Bversor = magnetic_field_->contravariant_versor(x_old, Btime);
-
-  // perform cartesian_boris step
-  IR3 v_new = electric_field_
-      ? boris_push(v_old, Oref_, Efield, Bmag, Bversor, dt)
-      : boris_push(v_old, Oref_, Bmag, Bversor, dt);
-  IR3 x_new = x_old + (Lref_ * dt) * v_new;
-
-  return {x_new[IR3::u], x_new[IR3::v], x_new[IR3::w],
-          v_new[IR3::u], v_new[IR3::v], v_new[IR3::w]};
+    const state& s, const double& time, const double& dt) const {
+  double Btime = time * iB_time_factor_;
+  const double tildeEref = Oref_ * iB_time_factor_ / (iE_time_factor_ * Vref_);
+
+  IR3 x = this->get_position(s), v = this->get_velocity(s);
+  double B = magnetic_field_->magnitude(x, Btime);
+  IR3 b = magnetic_field_->contravariant_versor(x, Btime);
+  IR3 E = electric_field_ ?
+      electric_field_->contravariant(x, time * iE_time_factor_) :
+      IR3 {0, 0, 0};
+
+  IR3 updated_v = electric_field_ ?
+      boris_push(v, Oref_, tildeEref, E, B, b, dt) :
+      boris_push(v, Oref_, B, b, dt);
+  IR3 updated_x = x + (Lref_ * dt) * updated_v;
+
+  return this->generate_state(updated_x, updated_v);
 }
 
 //! Returns the `cartesian_boris::state` state from a point in SI phase-space.
```

Comment:
I may have forgotten to update these comments. 
I think the state is no longer in SI units (at least not the velocity).
This update fixed this for the other steppers.

Reply:
As a rule (at least for the time being), cartesian coordinates means SI (i.e., meters). On the other hand, and as stated in the documentation, the velocity is normalised to Vref. Therefore, the phase-space is not in SI (as erroneously written), only the coordinates. This will be fixed.